src/main/java/cubicchunks/server/ServerCubeCache.java

```diff
@@ -307,79 +344,159 @@ public Cube getCube(int cubeX, int cubeY, int cubeZ, Requirement req) {
+			onCubeLoaded(cube, column);
 		}
 
-		if(cube == null) {
-			// generate the Cube
-			ICubePrimer primer = cubeGen.generateCube(cubeX, cubeY, cubeZ);
-			cube = new Cube(column, cubeY, primer);
+		return postProcessCube(cubeX, cubeY, cubeZ, cube, column, req);
+	}
 
+	/**
+	 * After successfully loading a cube, add it to it's column and the lookup table
+	 *
+	 * @param cube The cube that was loaded
+	 * @param column The column of the cube
+	 */
+	private void onCubeLoaded(@Nullable Cube cube, @Nonnull Column column) {
+		if(cube != null) {
 			column.addCube(cube);
 			cubemap.put(cube); // cache the Cube
-			this.worldServer.getFirstLightProcessor().initializeSkylight(cube); // init sky light, (does not require any other cubes, just OpacityIndex)
-			cube.onLoad(); // init the Cube
+			cube.onLoad();             // init the Cube
+		}
+	}
 
-			if(req.compareTo(Requirement.GENERATE) <= 0) {
+	/**
+	 * Process a recently loaded cube as per the specified effort level.
+	 *
+	 * @param cubeX Cube x position
+	 * @param cubeY Cube y position
+	 * @param cubeZ Cube z positon
+	 * @param cube The loaded cube, if loaded, else <code>null</code>
+	 * @param column The column of the cube
+	 * @param req Work done on the cube
+	 *
+	 * @return The processed cube, or <code>null</code> if the effort level is not sufficient to provide a cube
+	 */
+	@Nullable
+	private Cube postProcessCube(int cubeX, int cubeY, int cubeZ, @Nullable Cube cube, @Nonnull Column column, @Nonnull Requirement req) {
+		// Fast path - Nothing to do here
+		if (req == Requirement.LOAD) return cube;
+		if (req == Requirement.GENERATE && cube != null) return cube;
+
+		if(cube == null) {
+			// generate the Cube
+			cube = generateCube(cubeX, cubeY, cubeZ, column);
+			if(req == Requirement.GENERATE) {
 				return cube;
 			}
 		}
 
-		// forced full population of this Cube!
 		if(!cube.isFullyPopulated()) {
-			cubeGen.getPopulationRequirement(cube).forEachPoint((x, y, z) -> {
-				Cube popcube = getCube(x + cubeX, y + cubeY, z + cubeZ);
-				if(!popcube.isPopulated()) {
-					cubeGen.populate(popcube);
-					popcube.setPopulated(true);
-				}
-			});
-			cube.setFullyPopulated(true);
-		}
-		if(req == Requirement.POPULATE) {
-			return cube;
+			// forced full population of this cube
+			populateCube(cube);
+			if (req == Requirement.POPULATE) {
+				return cube;
+			}
 		}
 
 		//TODO: Direct skylight might have changed and even Cubes that have there
 		//      initial light done, there might be work to do for a cube that just loaded
 		if(!cube.isInitialLightingDone()) {
-			for(int x = -2;x <= 2;x++) {
-				for(int z = -2;z <= 2;z++) {
-					for(int y = 2;y >= -2;y--) {
-						if(x != 0 || y != 0 || z != 0) {
-							// FirstLightProcessor is so soft and fluffy that it can't even ask for Cubes correctly!
-							getCube(x + cubeX, y + cubeY, z + cubeZ);
-						}
+			initializeCubeSkylight(cube);
```

Comment:
This name is very similar to `FirstLightProcessor.initializeSkylight`, it's easy to confuse these 2. (that's the reason I initially called the `FirstLightprocessor` one `earlySkylightMap`, initialize suggests that after you call it - lighting is done)